apps/apps.c

```diff
@@ -2343,6 +2349,117 @@ int app_isdir(const char *name)
 }
 #endif
 
+/* app_dirname section */
+
+/*
+ * This exactly follows what POSIX's
+ * dirname does, but is implemented
+ * in a more platform independent way.
+ *
+ * path        dirname
+ * /usr/lib    /usr
+ * /usr/       /
+ * usr         .
+ * /           /
+ * .           .
+ * ..          .
+ * NULL        .
```

Comment:
NULL and . and pretty much anythign without a slash should be an error.  If someone types "usr" we have no idea if it's /usr or it's suppose to be a filename with that name.

Reply:
`dirname("usr")` in Unix semantics should return `.`, always.  Doesn't matter if it's a type `f` or a type `d` (in `find` speak)

Reply:
Apart from the NULL line, that table is a copy from `dirname(3)`, at least on my Linux box (that's where I got it from)

Reply:
sigh.  okay.  still question the need for this except on non-posix.

Reply:
and also question NULL.  we don't check for null params on our API :)

Reply:
Easily removed, will do

Reply:
I just realised that this comes from #3709 as well...  @InfoHunter?

Reply:
Yes, those were crafted by me in #3709... So the conclusion is:

a. change `fallback_dirname()` to `posix_dirname()`; and
b. remove support for NULL input?

And then you pull those new fixing commits from #3709 , @levitte ?

(And yes, that table was copied from `dirname(3)`...)

Reply:
Yes

Reply:
Commit has been updated in #3709 , please grab this: 605ca3d